[Merged by Bors] - feat: prove the Gagliardo-Nirenberg-Sobolev inequality

fpvandoorn opened the pull request:
* From the Sobolev inequality project

Co-authored-by: Heather Macbeth 25316162+hrmacbeth@users.noreply.github.com

---

- [x] depends on: #14124
- [x] depends on: #14159

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Ruben-VandeVelde commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,619 @@
+    ∫⁻ x, ∏ i, (∫⁻ xᵢ, f (update x i xᵢ) ∂μ i) ^ ((1 : ℝ) / (#ι - 1 : ℝ)) ∂.pi μ
+    ≤ (∫⁻ x, f x ∂.pi μ) ^ p := by
+  have h0 : (1:ℝ) < #ι := by norm_cast; exact Fintype.one_lt_card
+  have h1 : (0:ℝ) < #ι - 1 := by linarith
+  have h3 : 0 ≤ ((1 : ℝ) / (#ι - 1 : ℝ)) := by positivity
```
I guess there used to be a `h2` here; you might renumber

Ruben-VandeVelde commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,619 @@
+
+open FiniteDimensional
+
+set_option linter.unusedVariables false in
```
Why?

fpvandoorn replied:
It flags the hypotheses `hu` and `h2u` in the conclusion. But disabling the linter is not nice, I enabled it again.

Ruben-VandeVelde commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,619 @@
+  have hq : Real.IsConjExponent p q := .conjExponent hp
+  have h0p : p ≠ 0 := zero_lt_one.trans hp |>.ne'
+  have h1p : (p : ℝ) ≠ 1 := hq.one_lt.ne'
+  -- have h3p : (p : ℝ) ≠ 0 := hq.pos.ne'
```
```suggestion
```

Ruben-VandeVelde commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,619 @@
+  let C₁ : ℝ≥0 := ‖(e.symm : F' →L[ℝ] F)‖₊
+  let C₂ : ℝ≥0 := ‖(e : F →L[ℝ] F')‖₊
+  obtain ⟨C, hC⟩ := snorm_le_snorm_fderiv_of_eq_inner μ F' hp h2p hp'
+  refine ⟨C₁ * C * C₂, @fun u hu h2u ↦ ?_⟩ -- nope?
```
"nope?"?

fpvandoorn replied:
Oops, artifact from when the constant wasn't correct yet :-)

sgouezel commented on Mathlib/MeasureTheory/Function/LpSpace.lean:
```diff
@@ -672,6 +672,13 @@ theorem memℒp_indicator_iff_restrict (hs : MeasurableSet s) :
 
+/-- For a function `f` with support in `s`, the Lᵖ norms of `f` with respect to `μ` and
+`μ.restrict s` are the same. -/
+theorem snorm_restrict_eq {s : Set α} (hs : MeasurableSet s) (hsf : f.support ⊆ s) :
```
hs shouldn't be needed: otherwise, use the measurable hull of s instead. Since it contains s, it will also satisfy hsf

fpvandoorn replied:
I also thought that, but I wasn't sure how to get `snorm f p (μ.restrict s) = snorm f p (μ.restrict (toMeasurable μ s)`

sgouezel replied:
Loogle tells me about [MeasureTheory.Measure.restrict_toMeasurable_of_sFinite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Typeclasses.html#MeasureTheory.Measure.restrict_toMeasurable_of_sFinite)

fpvandoorn replied:
I replaced the lemma, and this is actually nicer. I didn't realize we had to put some effort in making the set measurable when we applied it, so this is a nice simplification.

This does add a small amount of imports to the file `LpSpace`.

sgouezel replied:
I have a proof removing the `SFinite` assumption, but I will first wait for #14260 to be merged, to avoid introducing new lemmas with a wrong name.

sgouezel replied:
See #14289

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,610 @@
+* `MeasureTheory.snorm_le_snorm_fderiv_of_le`:
+  The bound holds when `1 ≤ p < n`, `0 ≤ q` and `p⁻¹ - n⁻¹ ≤ q⁻¹`.
+  Note that in this case the constant depends on the support of `u`.
+-/
```
Could you sketch in the file-level docstring the proof you are following, first in informal terms and then describing the formalization concepts you use to avoid DTT?

fpvandoorn replied:
Done, and I also further gave a proof sketch in 5 long proofs.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,610 @@
+    (hp : (s.card : ℝ) * p ≤ 1)
+    (i : ι) (hi : i ∉ s) {f : (∀ i, A i) → ℝ≥0∞} (hf : Measurable f) :
+    T μ p f (insert i s) ≤ T μ p (∫⋯∫⁻_{i}, f ∂μ) s := by
+  calc T μ p f (insert i s)
```
Whenever a proof is more than 15 lines long, I think it deserves some comments explaining what you are doing. This one qualifies.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,610 @@
+    {u : (ι → ℝ) → F} (hu : ContDiff ℝ 1 u)
+    (h2u : HasCompactSupport u) :
+    ∫⁻ x, (‖u x‖₊ : ℝ≥0∞) ^ p ≤ (∫⁻ x, ‖fderiv ℝ u x‖₊) ^ p := by
+  have : (1:ℝ) ≤ ↑#ι - 1 := by
```
Add some comments in the proof?

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,610 @@
+
+/-! ## The Gagliardo-Nirenberg-Sobolev inequality -/
+
+variable [Nontrivial ι] {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F] [CompleteSpace F]
```
Do you really need the assumption that iota is nontrivial? Below, you assume `Real.IsConjExponent #ι p`, which probably implies it? Same thing in the other statements below.

fpvandoorn replied:
Well spotted! This removes multiple other arguments.
In `snorm_le_snorm_fderiv_of_eq_inner` I added a new condition that `0 < n`, and I *think* that is necessary, but my brain hurts when I have to think about 0-dimensional vector spaces.

fpvandoorn commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,610 @@
+-/
+theorem snorm_le_snorm_fderiv_of_eq_inner {p p' : ℝ≥0} (hp : 1 ≤ p)
+    (h2p : p < finrank ℝ E) (hp' : (p' : ℝ)⁻¹ = p⁻¹ - (finrank ℝ E : ℝ)⁻¹) :
+    ∃ C : ℝ≥0, ∀ {u : E → F'} (_hu : ContDiff ℝ 1 u) (_h2u : HasCompactSupport u),
```
Done. Note that in the versions with a finite dimensional normed space as codomain it does depend on (the norm of) `F`.

fpvandoorn commented:
I (hopefully) addressed all review comments.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,663 @@
+
+open FiniteDimensional
+
+variable (F) in
```
still needed?

sgouezel commented on Mathlib/MeasureTheory/Function/LpSpace.lean:
```diff
@@ -672,6 +673,15 @@ theorem memℒp_indicator_iff_restrict (hs : MeasurableSet s) :
 
+/-- For a function `f` with support in `s`, the Lᵖ norms of `f` with respect to `μ` and
+`μ.restrict s` are the same. -/
+theorem snorm_restrict_eq [SFinite μ] {s : Set α} (hsf : f.support ⊆ s) :
```
In fact I think it's possible to get rid also of the SFinite condition, but I'll do that after this PR is merged.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,661 @@
+
+/-! ## The Gagliardo-Nirenberg-Sobolev inequality -/
+
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F] [CompleteSpace F]
```
The completeness assumption here is probably not necessary, but removing this can definitely wait for another PR.

sgouezel replied:
Done in #14262

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,661 @@
+
+/-- The **Gagliardo-Nirenberg-Sobolev inequality**.  Let `u` be a continuously differentiable
+compactly-supported function `u` on a normed space `E` of finite dimension `n ≥ 2`, equipped
+with Haar measure. There exists a constant `C` depending only on `E`, such that the Lebesgue
```
```suggestion
with Haar measure. There exists a constant `C` depending only on `E` and `μ`, such that the Lebesgue
```

fpvandoorn replied:
I reformulated these sentences now that we've separated out the constants.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,661 @@
+integral of the pointwise expression `|u x| ^ (n / (n - 1))` is bounded above by `C` times the
+`n / (n - 1)`-th power of the Lebesgue integral of the Fréchet derivative of `u`. -/
+theorem lintegral_pow_le_pow_lintegral_fderiv {p : ℝ} (hp : Real.IsConjExponent (finrank ℝ E) p) :
+    ∃ C : ℝ≥0, ∀ {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F] [CompleteSpace F]
```
I have a small problem with existential statements like that, that the constant `C` might a priori depend on the universe of `F`. So you can not really write that "it only depend on E and mu" (well, it's true, but not what the statement says). I wonder if it's not clearer to define a constant `gagliardoNirenbergSobolevConstant E \mu := ...` and use that in the statements, to lift this ambiguity (with a shorter notation to be used in the current file, to make things readable). It would also have the advantage to avoid the ugly forall quantifier in the statement.

Same thing for the other statements below.

What do you think?

fpvandoorn replied:
I am in favor, since I think that will be a lot nicer to work with in practice. I'm also doing that in the formalization of Carleson's theorem. I wasn't sure if that style was appropriate for Mathlib.

fpvandoorn replied:
Since multiple statements in this file have slightly different constants, I'll probably use something like `lintegralPowLePowLintegralFderivConst`.

sgouezel reviewed (commented):
It feels weird to have this in the MeasureTheory folder; this is firmly a statement in analysis, just like all Sobolev spaces theory. Maybe create a folder `Analysis/FunctionalSpaces` and put it there?

fpvandoorn commented:
Thanks for your quick reviews! 
I don't see how to quickly do the 2 suggested generalizations, so let's leave them for a later PR.
I incorporated your last suggestion, reformulated docstrings with this change, and added a paragraph in the module doc explaining this.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,711 @@
+It only depends on `E`, `μ` and `p`.
+It is determined by the ratio of the measures on `E` and `ℝⁿ` and
+the operator norm of `e` (raised to suitable powers involving `p`).-/
+def lintegralPowLePowLIntegralFDerivConst (p : ℝ) : ℝ≥0 := by
```
`irreducible_def`, maybe, to make sure that Lean will not try to unfold it given how nasty it is. Same thing for all the other constants definition below.

sgouezel commented on Mathlib/MeasureTheory/Integral/SobolevInequality.lean:
```diff
@@ -0,0 +1,711 @@
+/-- The constant factor occurring in the conclusion of `lintegral_pow_le_pow_lintegral_fderiv`.
+It only depends on `E`, `μ` and `p`.
+It is determined by the ratio of the measures on `E` and `ℝⁿ` and
+the operator norm of `e` (raised to suitable powers involving `p`).-/
```
Can you say in the docstring what `e` is?

sgouezel reviewed (commented):
bors d+
Thanks!

fpvandoorn commented:
bors merge
